## 9. Функции

**Функции** – это многократно используемые фрагменты программы. Они позволяют дать имя определённому блоку команд с тем, чтобы в последствии запускать этот блок по указанному имени в любом месте программы и сколь угодно много раз. Это называется вызовом функции. Мы уже использовали много встроенных функций, как то len и range.

Функции определяются при помощи зарезервированного слова def. После этого слова указывается имя функции, за которым следует пара скобок, в которых можно указать имена некоторых переменных, и заключительное двоеточие в конце строки. Далее следует блок команд (инструкций), составляющих функцию.

**Сигнатура функции** — часть общего объявления функции, позволяющая средствам трансляции идентифицировать функцию среди других. 
Составляющие сигнатуры:
 
 1) имя функции;
 
 2) параметры функции;
 
 3) возвращаемые значения.

<img src="img/python-function-declaration.png" width=400 height=400 />


In [72]:
def say_hi():
    print('Hi!')
    print('Mr/Ms')

say_hi()

Hi!
Mr/Ms


## 9.1 Параметры функций

Функции могут принимать параметры, т.е. некоторые значения, передаваемые функции для того, чтобы она что-либо сделала с ними. Эти параметры похожи на переменные, за исключением того, что значение этих переменных указывается при вызове функции, и во время работы функции им уже присвоены их значения.

Параметры указываются в скобках при объявлении функции и разделяются запятыми. Аналогично мы передаём значения, когда вызываем функцию. Обратите внимание на терминологию: имена, указанные в объявлении функции, называются **параметрами**, тогда как значения, которые вы передаёте в функцию при её вызове, – **аргументами**.

In [73]:
def say_hi(name):
    name = name.capitalize()
    print(f'Hi, {name}!')

say_hi('Petr')
say_hi('masha')
say_hi(name='Petr')

Hi, Petr!
Hi, Masha!
Hi, Petr!


In [2]:
def print_max(a, b): 
    if a > b:
        print(a, 'is max') 
    elif a == b:
        print(a, 'equals to', b)
    else:
        print(b, 'is max')

print_max(6, 7)
print_max(3, 3)
print_max(a=3, b=3)
print_max(b=4, a=3)

x = 5
y = 2
print_max(x, y)

7 is max
3 equals to 3
3 equals to 3
4 is max
5 is max


In [3]:
def get_max_value(a: int, b: int) -> int | None: 
    if a > b:
        return a
    if a == b:
        return None
    return b

a = 5
b = 9
max_value = get_max_value(a, b)
if max_value is None:
    print(a, 'equals to', b)
else:
    print(max_value, 'is max')

9 is max


Подробнее о type hints можно посмотреть в документации:
https://docs.python.org/3/library/typing.html

## 9.2 Локальные переменные

При объявлении переменных внутри определения функции, они никоим образом не связаны с другими переменными с таким же именем за пределами функции – т.е. имена переменных являются локальными в функции. Это называется областью видимости переменной. **Область видимости всех переменных ограничена блоком**, в котором они объявлены, начиная с точки объявления имени.

In [76]:
x = 50

def func(x: int) -> int:
    x = 10
    my_value = 10
    print('x =', x)
    x = 2
    print('Replace x to', x)
    print('Y:', my_value)
    return x

print(f'Initial {x=}')
print('func(x) =', func(x))
print(f'Global {x=}')

Initial x=50
x = 10
Replace x to 2
Y: 10
func(x) = 2
Global x=50


In [79]:
x = 50

def func() -> None:
    print('x inside function =', x)

func()
print('x =', x)

x inside function = 50
x = 50


## 9.3 Global и как с ним жить

Чтобы присвоить некоторое значение переменной, определённой на высшем уровне программы (т.е. не в какой-либо области видимости, как то функции или классы), необходимо явно указать Python, что её имя не локально, а глобально (global). 

Без применения зарезервированного слова global невозможно присвоить значение переменной, определённой за пределами функции.

**P.S.** Избегайте использование global. Общее правило: если вы используете **global**, то вероятно что-то делаете не так. Это действительно необходимо в очень редких случаях.

In [82]:
x = 50

def func():
    global x
    
    print('x =', x)
    x = 2
    print('Replace x to', x)

func()
print('Result global x =', x)

x = 50
Replace x to 2
Result global x = 2


In [85]:
def func(x):
    x += 1
    return x
    
x = func(10)
print(x)

11


## 9.4 Nonlocal
**nonlocal** ограничивает область поиска областями видимости объемлющих функций; она требует, чтобы перечисленные в инструкции имена уже существовали, и позволяет присваивать им новые значения. В область поиска не входят глобальная и встроенная области видимости.

**global** вынуждает интерпретатор начинать поиск имен с области объемлющего  модуля  и  позволяет  присваивать  переменным  новые  значения.  Область поиска простирается вплоть до встроенной области видимости, если искомое имя не будет найдено в модуле, при этом операция присваивания значений глобальным  именам всегда  будет  создавать  или изменять переменные в области видимости модуля.


In [89]:
x = 10

def func():
    x = 2
    print('x =', x)
    
    def func_inner(): 
        def super_inner():
            nonlocal x
            x = 8
        super_inner()
        print('inner print', x)
        
    func_inner()
    print('Replace x to', x)
    
func()
print(x)

x = 2
inner print 8
Replace x to 8
10


In [8]:
def tester(start):
    state = start  # В каждом вызове сохраняется свое значение state
    def nested(label):        
        print(label, state) # в объемлющей области видимости
    return nested

f = tester(10)

f(1)
f(2)
f(3)

1 10
2 10
3 10


In [22]:
def tester(start):
    state = start  # В каждом вызове сохраняется свое значение state
    def nested(label):
        nonlocal state      # Объект state находится 

        print(label, state) # в объемлющей области видимости
        state += 1 # Изменит значение переменной, объявленной как nonlocal
    return nested

f = tester(0)
print(f)
f(1)
f(2)
f(3)

<function tester.<locals>.nested at 0x10361e5f0>
1 0
2 1
3 2


## 9.5 Значения аргументов по умолчанию

Зачастую часть параметров функций могут быть необязательными, и для них будут использоваться некоторые заданные значения по умолчанию, если пользователь не укажет собственных. Этого можно достичь с помощью значений аргументов по умолчанию. Их можно указать, добавив к имени параметра в определении функции оператор присваивания (=) с последующим значением.

**P.S.** Обратите внимание, что **значение по умолчанию** должно быть неизменяемым объектом. В очень редких случаях можно использовать изменяемые объекты.

In [91]:
# Задача: реализовать функцию, в которую имя может не передано
# в таком случае нужно использовать значение по умолчанию
def say_hi(name: str | None) -> None:
    if not name:
        name = 'Mr/Ms'
    name = name.capitalize()
    print(f'Hi, {name}!')

say_hi('Petr')
say_hi('masha')
say_hi(None)

Hi, Petr!
Hi, Masha!
Hi, Mr/ms!


In [95]:
# def say_hi(name='Mr/Ms'):
def say_hi(name: str = 'Mr/Ms'):
    name = name.capitalize()
    print(f'Hi, {name}!')
    
say_hi('Petr')
say_hi('masha')
say_hi()
# say_hi(None)

Hi, Petr!
Hi, Masha!
Hi, Mr/ms!


In [99]:
# def math_func(a, b, c, sigma) -> int:
#     return (a + b * c) / sigma

# def math_func(a, b, c) -> int:
#     return math_func(a, b, c, 1)

# def math_func(a, b) -> int:
#     return math_func(a, b, 10)

def math_func(a, b, c=10, sigma=1) -> int:
    return (a + b * c) / sigma


math_func(1, 2, 3)

7.0

In [97]:
# Перегрузки функций в Python нет, есть только переопределение
def math_func(a, b, c, sigma) -> int:
    return (a + b * c) / sigma

def math_func(a, b, c) -> int:
    return (a + b * c) / 1

def math_func(a, b) -> int:
    return (a + b * 10) / 1

help(math_func)

Help on function math_func in module __main__:

math_func(a, b) -> int



In [98]:
math_func = 1
math_func = 2
math_func = 3
print(math_func)

3


### В Python все есть объект. Даже функция?
Да!

In [101]:
id(max)

4407496192

In [100]:
type(max)

function

In [103]:
type(math_func)

function

In [104]:
# Еще раз будьте осторожны с переменными
max = 1
max([1, 2])

TypeError: 'int' object is not callable

In [106]:
def max(x: int, y: int) -> int:
    return x if x > y else y

max(1, 2)
max([1, 2])

TypeError: max() missing 1 required positional argument: 'y'

In [107]:
def max_values(x: float, y: float) -> float:
    return x if x > y else y

max_values(1.2, 2)

2

## 9.6 Именованные аргументы

Если имеется некоторая функция с большим числом параметров, и при её вызове требуется указать только некоторые из них, значения этих параметров могут задаваться по их имени – это называется ключевые (именованные) параметры. В этом случае для передачи аргументов функции используется имя (ключ) вместо позиции (как было до сих пор).

Есть два преимущества такого подхода: во-первых, использование функции становится легче, поскольку нет необходимости отслеживать порядок аргументов; во-вторых, можно задавать значения только некоторым избранным аргументам, при условии, что остальные параметры имеют значения аргумента по умолчанию.

In [113]:
def func(a, c=10, b=5):
    print('a =', a, ', b =', b, ', c =', c)

func(3, 7)
func(a=3, b=abs(7))
func(b=7, a=3)
func(25, c=24)
func(c=50, a=100)


a = 3 , b = 5 , c = 7
a = 3 , b = 7 , c = 10
a = 3 , b = 7 , c = 10
a = 25 , b = 5 , c = 24
a = 100 , b = 5 , c = 50


In [117]:
# Нельзя указывать параметр после параметра со значением по умолчанию
def func(c=10, a, b=5):
    print('a =', a, ', b =', b, ', c =', c)

# func(1) -- чему равны c, a, b ?
# func(1, 2, 3)
# func(2, 3)


SyntaxError: parameter without a default follows parameter with a default (3611634040.py, line 2)

In [112]:
# Можно явным образом заставить использовать именованные переменные
def func(*, a, c=10, b=5):
    print('a =', a, ', b =', b, ', c =', c)

func(a=2)
func(a=1, b=2, c=10)

# так не работает:
# func(1)

a = 2 , b = 5 , c = 10
a = 1 , b = 2 , c = 10


## 9.7 Переменное число параметров

In [119]:
print(1)
print(1, 2, 3)

1
1 2 3


In [124]:
def total_v1(initial: int = 0, *numbers: int, end: int) -> int: 
    count = initial
    print(numbers)

    for number in numbers: 
        count += number
    return count


print(total_v1(10, 1, 2, 3, 5, 6, 7, end=5))

# Опциональные параметры
def total_v2(initial=5, *args, **kwargs):
    return initial * 8

print(f'total_v2 = {total_v2(20)}')

# Именование переменных = успех
def total_v2(initial: int = 5, *numbers: int, **keywords: int) -> int: 
    count = initial
    print(f'{numbers=}')
    print(f'{keywords=}')
    for number in numbers: 
        count += number
    
    for value in keywords.values():
        count += value
    
    return count


print(total_v2(10, 1, 2, 3, vegetables=50, fruits=100))


# Усложняем
def total_v3(*numbers: int, initial: int = 10, apples: int = 10, **keywords: int): 
    print(f'{numbers=}')
    print(f'{keywords=}')
    count = initial
    for number in numbers: 
        count += number
    
    for value in keywords.values():
        count += value
    
    return count


total_v3(1, 2, 3, initial=0, apples=5, fruits=99)

(1, 2, 3, 5, 6, 7)
34
total_v2 = 160
numbers=(1, 2, 3)
keywords={'vegetables': 50, 'fruits': 100}
166
numbers=(1, 2, 3)
keywords={'fruits': 99}


105

In [125]:
# Опциональные аргументы, т.е. некоторые параметры не используются
def pow(x: int, power: int, *_):
    return x ** power

print(pow(3, 2, 8, 10, 'sdf'))

def get_person():
    return 'Ivan', 'Ivanov', 'Ivanovich'

# _ - можно использовать и для переменных
_, surname, _= get_person()

print(f'{surname=}')

9
surname='Ivanov'


In [50]:
def total(*numbers, extra_number, initial=5): 
    count = initial

    for number in numbers[:-1]: 
        count += number
    
    count += 2 * extra_number 
    print(count)

total(10, 1, 2, 3, extra_number=50)
total(1, 2, 3, extra_number=50, initial=10)


118
113


In [58]:
def custom_print(*args, **kwargs):
    print(*args, sep='|')
    
    for key, value in kwargs.items():
        print(f'{key} -> {value}')

custom_print('Hello')
custom_print('Hello', 'World')
custom_print('Hello', 'World', '!')

# Распаковка списков
elements = [1, 2, 3, 4]
custom_print(*elements) # *[1, 2, 3, 4] -> 1, 2, 3, 4

# Можно ли из словаря передавать именованные аргументы? -- Распаковка 
data = {'name': 'Petr', 'surname': 'Petrov'}
custom_print(name='Ivan', surname='Ivanov')

custom_print(**data) # **{'name': 'Petr', 'surname': 'Petrov'} -> name='Ivan', surname='Ivanov'

Hello
Hello|World
Hello|World|!
1|2|3|4

name -> Ivan
surname -> Ivanov

name -> Petr
surname -> Petrov


In [60]:
# Клонирование через распаковку
data = {'name': 'Petr', 'surname': 'Petrov'}
copy_dict = dict(**data)
print(id(data), id(copy_dict))

4408170048 4408176512


In [50]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



## 9.8 Оператор Return

Оператор return используется для возврата5 из функции, т.е. для прекращения её работы и выхода из неё. При этом можно также вернуть некоторое значение из функции.

In [62]:
# Type hint-ы очень полезны
def maximum(x, y): 
    if x > y:
        return x
    elif x == y:
        return 'Equals.' 
    else:
        return y

# Дополним type hint-ы, как вам?
def maximum(x: int, y: int) -> int | str: 
    if x > y:
        return x
    if x == y:
        return 'Equals.' 
    return y

# Так лучше
def maximum(x: int, y: int) -> int | None: 
    if x > y:
        return x
    if x == y:
        return None
    return y

def maximum(x: int, y: int) -> int: 
    if x > y:
        return x
    if x == y:
        return -1
    return y

print(maximum(2, 3))

3


In [65]:
# Заглушки - stub
# Функции без реализации
def stub():
    "abc"

def stub():
    pass


def stub():
    ...

# identical construction
def stub():
    Ellipsis


print(stub())

None


In [38]:
help(stub)

Help on function stub in module __main__:

stub()
    # identical construction



In [87]:
help(...)


Help on ellipsis object:

class ellipsis(object)
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __reduce__(...)
 |      Helper for pickle.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



## 9.9 Строки документации (docstring)

In [91]:
def print_max(a, b):
    """ 
    Prints max value of a and b.
    NOTE: prints equals if a = b.
    """

    if a > b:
        print(a, 'is max') 
    elif a == b:
        print(a, 'equals to', b)
    else:
        print(b, 'is max')


print_max(3, 5)

print(print_max.__doc__)

print('--------')
help(print_max)



5 is max
 
    Prints max value of a and b.
    NOTE: prints equals if a = b.
    
--------
Help on function print_max in module __main__:

print_max(a, b)
    Prints max value of a and b.
    NOTE: prints equals if a = b.

11


## 9.10 Анонимные функции (lambda)

Помимо инструкции def в языке Python имеется возможность создавать объекты функций в форме выражений. Т.е. простую функцию, у которой нет имени.

На практике они часто используются, как способ получить встроенную функцию или отложить выполнение фрагмента программного кода.

**lambda argument1, argument2,... argumentN : выражение, использующее аргументы**

<img src="img/python-lambda.png" width=400 height=400 />

In [70]:
def f(x: int, y: int) -> int:
    return x + y

# Anti-pattern -- почему?
f = lambda x, y: x + y 
f(1, 2)


3

In [98]:
array = [1, 2, 3, -19, 12]
# abs(1) >= abs(2)
# abs(2) >= abs(3)
# 1**2 >= 2**2..

def f(x):
    return x**2

max(array, key=abs)
max(array, key=f)
max(array, key=lambda x: x**2)
max(array, key=lambda x: x if x > 0 else x**2)
max(array, key=lambda x: x if x > 0 else -x)


-19

# Задачи

## 1. Список из числа.
 **Дано:** натуральное число N.
 
 **Задание:** написать функцию, которая возвращает список всех цифр этого числа в обратном порядке.
 
 **Пример:**
 
     123, результат: [3, 2, 1]

## 2. Палиндром.
 **Дано:**  слово, состоящее только из строчных латинских букв.
 
 **Задание:** написать функцию, которая возвращает True, если слово палиндромом, иначе False.
 
 Палиндро́м — число, буквосочетание, слово или текст, одинаково читающееся в обоих направлениях. Например, число 404; слова «топот» в русском языке; текст «а роза упала на лапу Азора» и пр.
 
 **Примеры:**
 
     'lol', результат: True
     'hi', результат: False
 
## 3. Деканат.
 **Дано:**  список студентов: каждый элемент списка содержит фамилию, имя, отчество, год рождения, курс, номер группы, оценки по пяти предметам.
 
 **Задание:** реализуйте сл. функции:

    1) возвращает список студентов по курсу, причем студенты одного курса располагались в алфавитном порядке;
    2) находит средний балл каждой группы по каждому предмету;
    3) определяет самого старшего студента и самого младшего студентов.
    4) возвращает словарь, где для каждой группы определен лучшый с точки зрения успеваемости студент.
 
  
## [Junior-] 4. Пешки.
 **Дано:** координаты расставленных пешек в виде набора строк.
 
 **Задание.** 
 Шахматы известны по всему миру, и практически всем людям знакомы их основные правила игры. В игре используется набор фигур, которые могут ходить по игровому полю различными способами, что обеспечивает огромное количество различных игровых комбинаций (к примеру, количество возможных шахматных партий оценивается Шенноном в 10^118). В этой задаче мы будем исследовать правила игры пешками.

Шахматы - это стратегическая игра двух игроков, которая разыгрывается на игровой доске с клетками, расположенными в восьми рядах (называемых горизонталями и обозначаемых цифрами от 1 до 8) и в восьми колонках (называемых вертикалями и обозначаемых буквами от a до h). Каждая клетка доски идентифицируется уникальной парой координат, состоящей из буквы и цифры (например, "a1", "h8", "d6"). В этой задаче мы будем иметь дело только с пешками. Пешка может бить пешку противника, которая находится перед ней в соседней клетке по диагонали справа или слева, переходя в эту клетку. У белых пешек клетки перед ними имеют номер горизонтали на единицу больше.

Сама по себе пешка является слабой фигурой, но мы можем использовать до восьми пешек для построения оборонительной стены. Стратегия оборонительной стены основывается на защите друг друга. Пешка защищена, если её клетка находится по ударом другой своей пешки. На игровом поле находятся только белые пешки. 
Вы должны реализовать функцию, позволяющую определить сколько пешек защищены в этой позиции.

Вам предоставляется набор координат, в которых расставлены белые пешки. Вы должны подсчитать количество защищенных пешек.
<img src='https://s3-eu-west-1.amazonaws.com/vitalypublicbucket/Screen+Shot+2017-05-04+at+14.23.57.png'/>

 **Примеры:**
 
     {"b4", "d4", "f4", "c3", "e3", "g5", "d2"}, результат: 6
     {"b4", "c4", "d4", "e4", "f4", "g4", "e5"}, результат: 1

## [Junior+] 5.Min-max.
 **Дано:** позиционный аргумент, как итерируемое или два или более позиционных аргументов. Необязательный ключевой аргумент, как функция.
 
 **Задание.** 
 
Роботы решили покопаться в себе и может быть даже улучшить что-нибудь. В этой задаче Вам нужно написать свою реализацию встроенных функций (версии для py3) min и max. Некоторые встроенные функции заблокированы здесь: import, eval, exec, globals. Не забудьте, что в этой задаче вам нужно реализовать две функции, а не одну как обычно.

max(iterable, *[, key]) or min(iterable, *[, key])
max(arg1, arg2, *args[, key]) or min(arg1, arg2, *args[, key])

Возвращает наибольший (наименьший) элемент в итерируемом (iterable) или наибольшее (наименьшее) из двух и более аргументов.

Если дан только один позиционный аргумент, то он должен быть итерируемым. В этом случае функция возвращает наибольший (наименьший) элемент из данного итерируемого. Если даны два или более позиционных аргумента, то возвращен будет наибольший (наименьший) из данных аргументов.

Необязательный ключевой аргумент key определяет функцию одного аргумента, которая используется для извлечения ключа для сравнения из каждого элемента массива (для примера, key=str.lower).

Если массив содержит несколько максимальных (минимальных) значений, то функция возвращает первый по порядку в массиве.

-- Python документация (Встроенные функции)

 
 **Примеры:**
 
    max(3, 2) == 3
    
    min(3, 2) == 2
    
    max([1, 2, 0, 3, 4]) == 4
    
    min("hello") == "e"
    
    max(2.2, 5.6, 5.9, key=int) == 5.6
    
    min([[1,2], [3, 4], [9, 0]], key=lambda x: x[1]) == [9, 0]
   